drivers/interrupt_controller/intc_miwu.c

```diff
@@ -0,0 +1,368 @@
+{
+	uint32_t base = DRV_CONFIG(miwu_devs[wui->table])->base;
+
+	NPCX_WKEN(base, wui->group) &= ~BIT(wui->bit);
+}
+
+unsigned int soc_miwu_irq_get_state(const struct npcx_wui *wui)
+{
+	uint32_t base = DRV_CONFIG(miwu_devs[wui->table])->base;
+
+	if (IS_BIT_SET(NPCX_WKEN(base, wui->group), wui->bit))
+		return 1;
+	else
+		return 0;
+}
+
+int soc_miwu_interrupt_configure(const struct npcx_wui *wui,
+		enum miwu_int_mode mode, enum miwu_int_trig trig)
+{
+	uint32_t base = DRV_CONFIG(miwu_devs[wui->table])->base;
+	uint8_t pmask = BIT(wui->bit);
+
+	if (mode == NPCX_MIWU_MODE_DISABLED) {
+		/* Clear MIWU enable bit */
+		NPCX_WKEN(base, wui->group) &= ~pmask;
+	} else {
+		/* Handle interrupt for level trigger */
+		if (mode == NPCX_MIWU_MODE_LEVEL) {
+			/* Set detection mode to level */
+			NPCX_WKMOD(base, wui->group) |= pmask;
+			switch (trig) {
+			/* Enable interrupting on level high */
+			case NPCX_MIWU_TRIG_HIGH:
+				NPCX_WKEDG(base, wui->group) &= ~pmask;
+				break;
+			/* Enable interrupting on level low */
+			case NPCX_MIWU_TRIG_LOW:
+				NPCX_WKEDG(base, wui->group) |= pmask;
+				break;
+			default:
+				return -EINVAL;
+			}
+		/* Handle interrupt for edge trigger */
+		} else {
+			/* Set detection mode to edge */
+			NPCX_WKMOD(base, wui->group) &= ~pmask;
+			switch (trig) {
+			/* Handle interrupting on falling edge */
+			case NPCX_MIWU_TRIG_LOW:
+				NPCX_WKAEDG(base, wui->group) &= ~pmask;
+				NPCX_WKEDG(base, wui->group) |= pmask;
+				break;
+			/* Handle interrupting on rising edge */
+			case NPCX_MIWU_TRIG_HIGH:
+				NPCX_WKAEDG(base, wui->group) &= ~pmask;
+				NPCX_WKEDG(base, wui->group) &= ~pmask;
+				break;
+			/* Handle interrupting on both edges */
+			case NPCX_MIWU_TRIG_BOTH:
+				/* Enable any edge */
+				NPCX_WKAEDG(base, wui->group) |= pmask;
+				break;
+			default:
+				return -EINVAL;
+			}
+		}
+
+		/* Enable wake-up input sources */
+		NPCX_WKINEN(base, wui->group) |= pmask;
+
+		/*
+		 * Clear pending bit since it might be set if WKINEN bit is
+		 * changed.
+		 */
+		NPCX_WKPCL(base, wui->group) |= pmask;
+	}
+
+	return 0;
+}
+
+void soc_miwu_init_gpio_callback(struct miwu_io_callback *callback,
+				const struct npcx_wui *io_wui, int port)
+{
+	/* Initialize WUI and GPIO settings in unused bits field */
+	callback->params.wui.table = io_wui->table;
+	callback->params.wui.group = io_wui->group;
+	callback->params.wui.bit   = io_wui->bit;
+	callback->params.gpio_port = port;
+}
+
+void soc_miwu_init_dev_callback(struct miwu_dev_callback *callback,
+				const struct npcx_wui *dev_wui,
+				miwu_dev_callback_handler_t handler,
+				struct device *source)
+{
+	/* Initialize WUI and input device settings */
+	callback->wui.table = dev_wui->table;
+	callback->wui.group = dev_wui->group;
+	callback->wui.bit   = dev_wui->bit;
+	callback->handler = handler;
+	callback->source = source;
+}
+
+int soc_miwu_manage_gpio_callback(struct miwu_io_callback *cb, bool set)
```

Comment:
Why can't we use `gpio_manage_callback` in gpio_utils.h here instead

Reply:
First, the parameter is different between soc_miwu_manage_gpio_callback() and gpio_manage_callback() although 'struct miwu_io_callback *cb' and struct gpio_callback *callback share the same memory space. In this PR, it won't increase code size since we don't call any gpio_manage_callback() and replace it with soc_miwu_manage_gpio_callback(). 

Of course, we can cast 'struct miwu_io_callback *'  back to 'struct gpio_callback * and call gpio_manage_callback() in MIWU driver directly. But I thought the current one is more straightforward.